### PR TITLE
feat(agents): Adds RLM harness: infinite context window (for some tasks)

### PR DESCRIPTION
**Pull request by cezarc1**
Adds a Recursive Language Model (RLM) engine.

Instead of dumping everything into the context window and hoping for the best (yolo), the model gets a sandboxed JS REPL where it can write code, see the output, and iterate. Think of it as giving the agent a notebook to work through problems step by step. The LLM model now uses the code to express logic vs. using it's own context window.

This is based on the [RLM paper](https://arxiv.org/pdf/2512.24601) and [DSPy's reference implementation](https://dspy.ai/api/modules/RLM/#when-to-use-rlm).

Fig2

## Why

For tasks over large context  (long documents, multi-file searches, information aggregation) RLMs consistently outperform standard LLM calls due to leveraging symbolic logic (read: code generation).

The key insight from the [paper](https://arxiv.org/pdf/2512.24601): the REPL becomes the model's working memory instead of the context window, so the model can strategically explore and accumulate results without context degradation.

See results:
rlm

## Scope

- **RLM harness** (`harness-rlm.ts`) — Sandboxed VM loop with bounded iterations, depth-limited recursion, object store for intermediates, retry logic, and extract-fallback when iterations run out.
- **Runtime API** — `context_read/search`, `repo_read/search`, `llm_query`, `tool_call`, `get_var/set_var`, `print`, `submit`.
- **REPL prompt aligned with DSPy** — Explore first, iterate small, verify before submitting, use llm_query for semantics, submit only after seeing outputs.
- **`rlm_call` agent tool** + **`/rlm` slash command** — programmatic and interactive entry points.
- **Config** — `tools.rlm.*` with tunables for depth, iterations, LLM budget, timeout.

## How to enable/use

Disabled by default. Flip one flag:

```yaml
tools:
  rlm:
    enabled: true
```

Then: `/rlm summarize this project` or let the agent use `rlm_call` on its own.

Optional tunables: `maxDepth` (0–8), `maxIterations` (1–96), `maxLlmCalls` (1–2048), `timeoutSeconds`, `extractOnMaxIterations`.

<h3>Greptile Summary</h3>

Adds RLM (Recursive Language Model) harness that gives agents a sandboxed JavaScript REPL to iteratively solve tasks requiring large context exploration. Instead of dumping everything into the context window, the model writes small code snippets, observes outputs, and uses symbolic retrieval/slicing to work through problems step-by-step. The REPL becomes the working memory instead of the context window.

**Major changes:**
- Core RLM harness (`harness-rlm.ts`) with bounded iterations, depth-limited recursion, object store for intermediates, retry logic, and extract-fallback when iterations run out
- Runtime API: `context_read/search`, `repo_read/search`, `llm_query`, `tool_call`, `get_var/set_var`, `print`, `submit`
- `rlm_call` agent tool and `/rlm` slash command for programmatic and interactive entry points
- Config schema (`tools.rlm.*`) with tunables for depth, iterations, LLM budget, timeout
- Disabled by default; requires explicit `tools.rlm.enabled: true`

**Implementation quality:**
- Comprehensive test coverage including behavior tests for edge cases (unhandled rejections, Promise handling, retries)
- Proper resource cleanup with tmpdir management
- Good error handling with retry logic for transient failures
- Security-conscious: workspace path validation, file size limits, binary detection, rate limiting
- Well-documented constants and clear separation of concerns

<h3>Confidence Score: 4/5</h3>

- This PR is safe to merge with minor considerations
- The implementation is well-designed with comprehensive tests, proper resource management, and security controls. The RLM harness includes workspace path validation, file size limits, recursion depth limits, and timeout enforcement. The feature is disabled by default and opt-in. The VM sandbox limitation is acknowledged and acceptable given the threat model. Code follows repository conventions and includes extensive behavioral tests for edge cases. Minor attention needed around VM context isolation and ensuring tool policy integration is complete.
- Pay close attention to `src/commands/agent/harness-rlm.ts` VM sandbox implementation (lines 1675-1695) to ensure the limited context is sufficient and tool access boundaries are properly enforced

<sub>Last reviewed commit: eb81c9a</sub>

<sub>(1/5) You can manually trigger the agent by mentioning @greptileai in a comment!</sub>

### CODE REVIEW

File: src/commands/agent/harness-rlm.ts [hunk 1]
```diff
@@ -0,0 +1,1831 @@
...
+      const pendingSubmitOps: Promise<unknown>[] = [];
+      let toolCallsInStep = 0;
+
+      const trackStepOp = <T>(op: Promise<T>): Promise<T> => {
+        // REPL code can forget to await async runtime calls; track them so we can drain before
+        // tmpRoot cleanup and avoid unhandled rejections.
+        void op.catch(() => undefined);
+        pendingStepOps.push(op);
+        return op;
+      };
```
[the remark this answers is not in the document]
Comment (cezarc1) [reply src/commands/agent/harness-rlm.ts:1234]: done

File: src/commands/agent/harness-rlm.ts [hunk 2]
```diff
@@ -0,0 +1,1831 @@
...
+  const root = await fs
+    .realpath(path.resolve(workspaceRoot))
+    .catch(() => path.resolve(workspaceRoot));
+  const candidate = path.isAbsolute(raw) ? path.resolve(raw) : path.resolve(root, raw);
+  const real = await fs.realpath(candidate);
+  const prefix = `${root}${path.sep}`;
+  if (real !== root && !real.startsWith(prefix)) {
+    throw new Error("path resolves outside workspace root");
+  }
+  return real;
```
[the remark this answers is not in the document]
Comment (cezarc1) [reply src/commands/agent/harness-rlm.ts:658]: done.

File: src/commands/agent/harness-rlm.ts [hunk 3]
```diff
@@ -0,0 +1,1831 @@
...
+      let scriptError: unknown;
+      let returnValue: unknown;
+      try {
+        const script = new vm.Script(wrappedCode, {
+          filename: `rlm-d${solveCtx.depth}-i${iteration}.js`,
+        });
+        // NOTE: Node.js vm is NOT a security sandbox. The model already has tool
+        // access (exec, file I/O) in normal operation so this does not expand the
+        // trust boundary. The vm context prevents accidental Node.js API use and
+        // enforces a synchronous timeout. RLM is gated behind tools.rlm.enabled.
```
[the remark this answers is not in the document]
Comment (cezarc1) [reply src/commands/agent/harness-rlm.ts]: done

File: src/commands/agent/harness-rlm.ts [hunk 4]
```diff
@@ -0,0 +1,1831 @@
...
+        groupSpace: params.groupSpace,
+        currentChannelId: params.currentChannelId,
+        currentThreadTs: params.currentThreadTs,
+        replyToMode: params.replyToMode,
+        hasRepliedRef: params.hasRepliedRef,
+        abortSignal: params.abortSignal,
+        senderIsOwner: params.senderIsOwner,
+      }),
+    ) as RlmToolDefinition[];
+    return toolDefs.filter((tool) => tool.name !== "rlm_call");
```
[the remark this answers is not in the document]
Comment (cezarc1) [reply src/commands/agent/harness-rlm.ts:711]: done

**Comment by cezarc1** [comment]
this PR should be ready for review. This formating failure is not from this PR: https://github.com/openclaw/openclaw/actions/runs/22084374452/job/63815790106?pr=17749

**Comment by obviyus** [comment]
This is super interesting! I'll take a look.
